### PR TITLE
proofs: refactor into multiple functions

### PR DESCRIPTION
**Pull request by gabriele-0201**

### CODE REVIEW

**Comment by gabriele-0201** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/208?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#208** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/208?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#155** <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/155?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#156](https://github.com/thrumdev/nomt/pull/156) <a href="https://app.graphite.dev/github/pr/thrumdev/nomt/156?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @gabriele-0201 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: core/src/multi_proof.rs [hunk 1]
```diff
@@ -42,6 +42,106 @@ pub struct MultiProof {
...
+        left: PathProofRange,
+        right: PathProofRange,
+    },
+    Advance {
+        sibling: Node,
+    },
+}
+
+impl PathProofRange {
+    fn prove_unique_path_remainder(&self, path_proofs: &Vec<PathProof>) -> Option<SubPathProof> {
```
Comment (rphmeier) [inline core/src/multi_proof.rs:69]: never use &Vec

File: core/src/multi_proof.rs [hunk 2]
```diff
@@ -42,6 +42,106 @@ pub struct MultiProof {
...
+        // If PathProofRange contains only one path_proofs
+        // then all remaining siblings are required for the multiproof
+
+        if self.lower != self.upper - 1 {
+            return None;
```
Comment (rphmeier) [inline core/src/multi_proof.rs:74]: fail fast (panic) if this function requires being called with a single proof range.

**Review by rphmeier — APPROVED** [review]

File: core/src/multi_proof.rs (hunk 1 above)
Comment (gabriele-0201) [reply core/src/multi_proof.rs]: yeah, ok, thanks!

Are there other reasons not to use `&Vec` beside restricting the caller to construct a vec if they just have a slice?

**Pull request merged**